DMCompiler/DM/Visitors/DMProcBuilder.cs

```diff
@@ -412,8 +412,33 @@ public void ProcessStatementSet(DMASTProcStatementSet statementSet)
         }
 
         public void ProcessStatementForList(DMASTProcStatementForList statementForList) {
-            DMExpression.Emit(_dmObject, _proc, statementForList.List);
-            _proc.CreateListEnumerator();
+
+            DMASTProcStatementVarDeclaration varDeclaration = statementForList.Initializer as DMASTProcStatementVarDeclaration;
+
+            if (varDeclaration?.Type != null)
+            {
+                // Atoms & world will use their contents list implicitly
+                // This is for things like "for(var/client/C)"
+                if (statementForList.List is null && !DMObjectTree.GetDMObject(varDeclaration.Type.Value).IsSubtypeOf(DreamPath.Atom))
+                {
+                    _proc.PushPath(varDeclaration.Type.Value);
+                    _proc.CreateTypeEnumerator();
+                }
+                else
+                {
+                    // If List is null it's "for(var/atom/A)" (or an atom subtype) which is implicitly "in world"
+                    DMExpression.Emit(_dmObject, _proc, statementForList.List ?? new DMASTIdentifier(statementForList.Location, "world"));
+                    _proc.PushPath(varDeclaration.Type.Value);
+                    _proc.CreateListEnumerator();
+                }
```

Comment:
It would be better to always use `CreateTypeEnumerator` and have the runtime decide to use the world's contents if you're enumerating an atom type.